theano/contrib/mkl/basic_ops.py

```diff
@@ -0,0 +1,472 @@
+    def make_node(self, x, gz):
+        out = x.type()
+        return Apply(self, [x, gz], [out])
+
+    def c_code(self, node, name, inp, out, sub):
+        x, gz, = inp
+        z, = out
+        sub['x'] = x
+        sub['gz'] = gz
+        sub['z'] = z
+        sub['name'] = U2IGrad.__name__
+        if 'float32' == node.inputs[0].type.dtype:
+            sub['precision'] = "F32"
+            sub['x_item_size'] = 4
+        elif "float64" == node.inputs[0].type.dtype:
+            sub['precision'] = "F64"
+            sub['x_item_size'] = 8
+        else:
+            raise TypeError("Type %s not implemented" %
+                            node.inputs[0].type.dtype)
+        ccode = """
+        if(NULL == %(z)s) {
+            npy_intp dims[4] = {0};
+            dims[0] = MKLNdarray_DIMS(%(gz)s)[0];
+            dims[1] = MKLNdarray_DIMS(%(gz)s)[1];
+            dims[2] = MKLNdarray_DIMS(%(gz)s)[2];
+            dims[3] = MKLNdarray_DIMS(%(gz)s)[3];
+
+            %(z)s = (PyArrayObject*)PyArray_ZEROS(MKLNdarray_NDIM(%(gz)s),
+                                                  dims,
+                                                  MKLNdarray_TYPE(%(gz)s),
+                                                  0);
+            if(NULL == %(z)s) {
+                %(fail)s;
+            }
+        }
+
+        //create usr layerout
+        if (1 == first_run) {
+            size_t z_size[4] = {0};
+            size_t z_stride[4] = {0};
+            int ndim = (int)MKLNdarray_NDIM(%(gz)s);
+            assert(ndim == DIMENSION);
+
+            for(int i=0; i<DIMENSION; i++) {
+                z_size[i] = (size_t)PyArray_DIMS(%(z)s)[ndim-i-1];
+                z_stride[i] = (size_t)PyArray_STRIDES(%(z)s)[ndim-i-1] / %(x_item_size)s;
+            }
+
+            CHECK_ERR( dnnLayoutCreate_%(precision)s(&layout_user,
+                                                     ndim,
+                                                     z_size,
+                                                     z_stride), err);
+
+            if (!dnnLayoutCompare_%(precision)s(MKLNdarray_LAYOUT(%(gz)s), layout_user) {
+                CHECK_ERR( dnnConversionCreate_%(precision)s(&from_internal,
+                                                             MKLNdarray_LAYOUT(%(gz)s),
+                                                             layout_user), err);
+            } else {
+                from_internal = NULL;
+            }
+        }
+
+        if (from_internal) {
+            CHECK_ERR( dnnConversionExecute_%(precision)s(from_internal,
+                                                          MKLNdarray_DATA(%(gz)s),
+                                                          PyArray_DATA(%(z)s)), err);
+        } else {
+            memcpy ((void*)PyArray_DATA(%(z)s), MKLNdarray_DATA(%(gz)s), PyArray_SIZE(%(z)s)*PyArray_ITEMSIZE(%(z)s));
+        }
+        first_run = 0;
+        """ % sub
+        return ccode
+
+
+class I2UGrad(BaseConvertOp):
```

Comment:
In the PR comments, you told it is equivalent to I2U. So why do need this one? What is the difference? Can you merge the 2 ops?

Here, gz should have the same size as x. So you should be able to have only 1 inputs to this op.